src/fabric/src/fabric_view_row.erl

```diff
@@ -0,0 +1,494 @@
+                #{},
+                Props
+            ),
+            {view_row, Row};
+        false ->
+            #view_row{
+                key = couch_util:get_value(key, Props),
+                id = couch_util:get_value(id, Props),
+                value = couch_util:get_value(value, Props),
+                doc = couch_util:get_value(doc, Props),
+                worker = couch_util:get_value(worker, Props)
+            }
+    end.
+
+get_id(#view_row{id = Id}) ->
+    Id;
+get_id({view_row, #{id := Id}}) ->
+    Id;
+get_id({view_row, #{}}) ->
+    undefined.
+
+get_key(#view_row{key = Key}) ->
+    Key;
+get_key({view_row, #{key := Key}}) ->
+    Key;
+get_key({view_row, #{}}) ->
+    undefined.
+
+set_key(#view_row{} = Row, Key) ->
+    Row#view_row{key = Key};
+set_key({view_row, #{} = Row}, Key) ->
+    {view_row, Row#{key => Key}}.
+
+get_value(#view_row{value = Value}) ->
+    Value;
+get_value({view_row, #{value := Value}}) ->
+    Value;
+get_value({view_row, #{}}) ->
+    undefined.
+
+get_doc(#view_row{doc = Doc}) ->
+    Doc;
+get_doc({view_row, #{doc := Doc}}) ->
+    Doc;
+get_doc({view_row, #{}}) ->
+    undefined.
+
+set_doc(#view_row{} = Row, Doc) ->
+    Row#view_row{doc = Doc};
+set_doc({view_row, #{} = Row}, Doc) ->
+    {view_row, Row#{doc => Doc}}.
+
+get_worker(#view_row{worker = Worker}) ->
+    Worker;
+get_worker({view_row, #{worker := Worker}}) ->
+    Worker;
+get_worker({view_row, #{}}) ->
+    undefined.
+
+set_worker(#view_row{} = Row, Worker) ->
+    Row#view_row{worker = Worker};
+set_worker({view_row, #{} = Row}, Worker) ->
+    {view_row, Row#{worker => Worker}}.
+
+transform(#view_row{value = {[{reduce_overflow_error, Msg}]}}) ->
+    {row, [{key, null}, {id, error}, {value, reduce_overflow_error}, {reason, Msg}]};
+transform(#view_row{key = Key, id = reduced, value = Value}) ->
+    {row, [{key, Key}, {value, Value}]};
+transform(#view_row{key = Key, id = undefined}) ->
+    {row, [{key, Key}, {id, error}, {value, not_found}]};
+transform(#view_row{key = Key, id = Id, value = Value, doc = undefined}) ->
+    {row, [{id, Id}, {key, Key}, {value, Value}]};
+transform(#view_row{key = Key, doc = {error, Reason}}) ->
+    {row, [{id, error}, {key, Key}, {value, Reason}]};
+transform(#view_row{key = Key, id = Id, value = Value, doc = Doc}) ->
+    {row, [{id, Id}, {key, Key}, {value, Value}, {doc, Doc}]};
+transform({view_row, #{} = Row0}) ->
+    Id = maps:get(id, Row0, undefined),
+    Key = maps:get(key, Row0, undefined),
+    Value = maps:get(value, Row0, undefined),
+    Doc = maps:get(doc, Row0, undefined),
+    Worker = maps:get(worker, Row0, undefined),
+    Row = #view_row{id = Id, key = Key, value = Value, doc = Doc, worker = Worker},
+    transform(Row).
+
+merge(Dir, Row, Rows) ->
+    lists:merge(
+        fun(RowA, RowB) ->
+            IdA = get_id(RowA),
+            IdB = get_id(RowB),
+            case Dir of
+                fwd -> IdA =< IdB;
+                rev -> IdA > IdB
```

Comment:
Hmm, I wonder if it should be `A<B` and `A>B`?

Erlang lists module expect the comparison operator to basically be `=<` or `>=`. However our btrees use `less` (strictly `<`). I don't know if that's an issue here since we don't expect two equal document IDs to come through once we start streaming for all_docs case. But it might be good to stick to `<` and `> `just to emphasize that
think our sorting function on the btrees is a strictly `less than` function and we want to make sure we use the same sorting function in the workers as we do in the coordinator.

Reply:
I have no either deeper understanding of this part or strong preferences, so I am fine with the suggestion.